### DISCUSSION TITLE
next.config.js rewrites and page refresh

### DISCUSSION THREAD

**Discussion by tar-aldev** [vercel/next.js#36657] [Help]
Hi. 
I have a small question about rewrites and page reload. Let me give the context.
We have next app and usual react app.

Next app has some pages and bigger part is in react app.
Currently, if we need to go to react app from next.js app we trigger the page reload, otherwise it was not working some time ago.

For now I am able to make it work with `router.push()`. So assume `<Link />` component from next.js should also work.
But still I see that the page is refreshed.
So let give an example.

1. We have a products page and each product has add to cart button. This page is a part of **next.js** app
2. we click on add to cart button on this next.js page
3. when add to cart button is clicked a new page opens, which is a part of **react app**.
4. the page is reloaded at this point and page from react app is shown.

Our next.config.js rewrites part looks llke this
```
{
          source: '/:path*',
          destination: `http://localhost:4200/:path*`
}
```

And actually, for react part it is the same - if the route is not part of react app, we have to force the page reload (using a `window.location` or common `<a>` instead of React router `<Link>`)

So, my question is - is there any way to get rid of page refresh?
Sorry I am not really familiar with config overrides, so any help is really appreciated!

**Comment by icyJoseph** [comment]
I am not sure I follow the entire question/problem, but I've set up just now, a Vite project using Client Side rendering, with SPA routed by react-router-dom, running at port 3001, and a Next.js App., at port 3000.

I tell `next.config.js` to do this:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  async rewrites() {
    return [
      {
        source: "/:path*",
        destination: "/:path*",
      },
      {
        source: "/:path*",
        destination: "http://localhost:3001/:path*", // Rewrite non-migrated routes to Create React App
      },
    ];
  },
};

module.exports = nextConfig;
```

Then from `next.js` I can reach the `blog` route on Vite.

A couple of observations.

- From next.js, clicking a link to `vite` will cause a navigation event, unlike SPA's, where the there's a route transition.
![Screenshot 2022-05-03 at 20 54 25](https://user-images.githubusercontent.com/21013447/166522996-561ec65b-556e-49fc-ab68-6cac3697daab.png)
- The CSP app, doesn't really have a `blog.html` file, there's just an index.html, so everything has to be piped through the same entry point
- The next.js app React tree disappears on the rewrite transition
- The URL shown to the user is maintained (the next.js URL)
- Notice that it is possible for my Next.js App to be at React version x.xx while the CSP Vite app can have a totally different version
- You can see in the React Dev Tools that the entire Client side React tree made by Next.js is gone! 
- Clicking the back button from

Because of the rewrites, being handled by a server, going from Next.js to CSP Vite, can be done with a relative link using `next/link`, but once the user receives the CSP Vite app, however, when the CSP Vite App wants to navigate to a Next.js page, it has to do it with an anchor tag using absolute paths. 

This is because the way SPA's work is a little deceiving, you need to use the Link component from the framework, because they prevent the default navigation behaviour, update the URL shown to the user and render the Routes that match the current path, while keeping track of this in their internal state.

Have I covered your doubts more or less?

**Reply by tar-aldev** [reply]
Indeed.
So from what I see, it is not possible to avoid page reloads right?
So to keep kind of SPA experience, even when navigating between the apps.

I can understand why it works this way. I wonder if may it is possible to bundle react app then and serve it as static files with custom next.js server?

So, the main concern is that going from one app to another is causing page reload.

**Reply by icyJoseph** [reply]
Perhaps, I have never done that, but it looks like migrating the non-next.js part to next.js would be easier.

**Reply by tar-aldev** [reply]
So did I get it right that, the best option would be to just use next.js for entire app?
And if we have some private pages, that should not be optimised for SEO in any way, we could just treat these pages as common react components and do all operations client side?

**Reply by icyJoseph** [reply]
Yeah, I mean, Next.js is offers the best of both worlds. 

If you have a client page that needs a tons of data you can pre-render the things that are common to all users and fetch the data from the client, or use `getServerSideProps` and generate the data for that user on the server. 

You could even use dynamic components and such.

You can hide these private pages from bots with <meta> tags, something you can't really set in CSP. There's even a library for that next-seo.

Short answer yeah. Next.js is more than just SSR, it offers for the most part, all possible models.

**Reply by GabenGar** [reply]
Even without the server stuff you can use `next export` and move your client react pages to static nextjs pages and then decide the amount of SEO/protection needed on per-page basis later.

**Reply by icyJoseph** [reply]
+1 to @GabenGar, I always forget about export, though I maintain 2 projects that use it 😅
